internal/terraform/evaluate.go

```diff
@@ -268,11 +268,27 @@ func (d *evaluationStateData) GetInputVariable(addr addrs.InputVariable, rng tfd
 	}
 
 	val, isSet := vals[addr.Name]
-	if !isSet {
-		if config.Default != cty.NilVal {
-			return config.Default, diags
-		}
-		return cty.UnknownVal(config.Type), diags
+	switch {
+	case !isSet:
+		// The config loader will ensure there is a default if the value is not
+		// set at all.
+		val = config.Default
+
+	case val.IsNull() && !config.Nullable && config.Default != cty.NilVal:
+		// If nullable=false a null value will use the configured default.
+		val = config.Default
+
+	case val.IsNull() && !config.Nullable:
+		// The value cannot be null, and there is no configured default.
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  `Invalid variable value`,
+			Detail:   fmt.Sprintf(`The resolved value of variable %q cannot be null.`, addr.Name),
+			Subject:  &config.DeclRange,
+		})
```

Comment:
It feels a bit funny to be catching this definition-related error in here, since the `evaluationStateData` component is normally concerned with validating references to other objects rather than validating those objects themselves. That is, in source range terms we're validating what's in `rng` rather than what's in `config.DeclRange` here.

Is there a "validate input variable" graph node we could check this in instead, so that we can be sure to always catch it even if there aren't any references to the variable? (Perhaps we always end up in here even if there aren't any references as a side-effect of some other quirky decision anyway, but if so I'd rather not rely on that, since if I were looking for this validation rule I'd expect to find it somewhere else. :thinking: )

Reply:
Ah, you're right. This ended up here because that's where I was working to assign the default, but it could be could be done earlier in evaluation, since we have the variable configuration at the point of module call evaluation. 
We use the same variable node for all actions (which is another issue), but I can move this validation check into there.

Reply:
On a related note, I'd suggest making this message say something more like "this variable is required", which is what we'd say if someone tried to assign a `null` to a provider argument whose schema says `Required: true`. I assume we already have some error text like this for the situation of omitting a required variable altogether, so maybe we can reuse that either in its entirety or with some adjustments.

Reply:
The reason I chose the wording here is that the biggest stumbling point for users has been when assigning variables in a module expanded with `for_each`, and I though it would be less confusing to state that it cannot be `null` since they _are_ assigning a value, it just requires a non-null value to be assigned and one of the `each.value` may be null. The error from the provider is a good point though, and perhaps ensuring they match is more important for now.

Reply:
Moved the validation check into the module call evaluation, and created what I think is a better error message, indicating that the variable is required, but the given value evaluated to null.